cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -0,0 +1,239 @@
+    # https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/resolvers/exotics/gitlab-resolver.js#L10C10-L10C23
+    if parsed_url.path.endswith((".tar", ".tar.gz", ".tgz")):
+        return True
+
+    # https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/resolvers/exotics/github-resolver.js#L24
+    if parsed_url.hostname == "codeload.github.com" and "tar.gz" in parsed_url.path:
+        return True
+
+    return False
+
+
+def _is_git_url(url: str) -> bool:
+    """Return True if a package URL is a git URL."""
+    # https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/resolvers/exotics/git-resolver.js#L32
+    if any(matcher.match(url) for matcher in GIT_PATTERN_MATCHERS):
+        return True
+
+    # https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/resolvers/exotics/git-resolver.js#L39
+    parsed_url = urlparse(url)
+    if parsed_url.hostname in GIT_HOSTS:
+        path_segments = [segment for segment in parsed_url.path.split("/") if segment]
+        # Return True if the path has exactly two segments (e.g. org/repo, not org/repo/file.tar.gz)
+        return len(path_segments) == 2
+
+    return False
+
+
+def _is_from_npm_registry(url: str) -> bool:
+    """Return True if a package URL is from the NPM or Yarn registry."""
+    return urlparse(url).hostname in NPM_REGISTRY_CNAMES
+
+
+def _get_packages_from_lockfile(source_dir: RootedPath) -> list[YarnClassicPackage]:
+    """Return a list of Packages for all dependencies in yarn.lock."""
+    lockfile_path = source_dir.join_within_root("yarn.lock")
+    lockfile = Lockfile.from_file(lockfile_path)
+    pyarn_packages: list[PYarnPackage] = lockfile.packages()
+    package_factory = _YarnClassicPackageFactory(source_dir)
+
+    return [
+        package_factory.create_package_from_pyarn_package(package) for package in pyarn_packages
+    ]
+
+
+def _get_main_package(source_dir: RootedPath) -> WorkspacePackage:
+    """Return a WorkspacePackage for the main package in package.json."""
+    # TODO: Use yarn classic project files once available
+    package_json = PackageJson.from_file(source_dir.join_within_root("package.json"))
+    if "name" not in package_json._data:
+        raise PackageRejected(
+            f"The package.json file located at {package_json._path.path} is missing the name field",
+            solution="Ensure the package.json file has a valid name.",
+        )
+    return WorkspacePackage(
+        name=package_json._data["name"],
+        version=package_json._data.get("version"),
+        relpath=package_json._path.subpath_from_root.parent,
+    )
+
+
+def _get_workspace_packages(
+    source_dir: RootedPath, workspaces: list[Workspace]
+) -> list[WorkspacePackage]:
+    """Return a WorkspacePackage for each Workspace."""
+    package_factory = _YarnClassicPackageFactory(source_dir)
```

Comment:
non-blocking nitpick: it bothered me a little that we need to instantiate `_YarnClassicPackageFactory` twice. Looking at the `create_package_from_workspace` function, it seems to be pretty straightforward and the only instance data it needs is `source_dir.path`.

Do you think we could simply perform the `WorkspacePackage` instantiation within this function here, instead of having it as a `_YarnClassicPackageFactory` member function?

Reply:
Sure, done :+1: